### ISSUE TITLE
why every predict results are different?

### ISSUE THREAD

**Issue by showkeyjar** [vturrisi/solo-learn#214]
I use `main_pretrain.py` to train custom data,  
but when I copy predict code from `main_linear.py` and predict image
every time the result is different.
for example:
first time: [0,0,0,0,0,2]
second time: [4,4,4,4,4,4]
……
this result make me surprise.
I think maybe I miss some random operation, but found nothing in code.
would you please help me how to solve this problem?

this is my train bash:
```
python solo-learn/main_pretrain.py \
    --dataset custom \
    --backbone resnet18 \
    --data_dir /mnt/data/images \
    --train_dir cheek \
    --no_labels \
    --max_epochs 400 \
    --gpus 0,1,2,3 \
    --accelerator gpu \
    --strategy ddp \
    --sync_batchnorm \
    --precision 16 \
    --optimizer sgd \
    --lars \
    --grad_clip_lars \
    --eta_lars 0.02 \
    --exclude_bias_n_norm \
    --scheduler warmup_cosine \
    --lr 1.0 \
    --classifier_lr 0.1 \
    --weight_decay 1e-5 \
    --batch_size 128 \
    --num_workers 4 \
    --brightness 0.4 \
    --contrast 0.4 \
    --saturation 0.2 \
    --hue 0.1 \
    --color_jitter_prob 0.8 \
    --gray_scale_prob 0.2 \
    --horizontal_flip_prob 0.5 \
    --gaussian_prob 1.0 0.1 \
    --solarization_prob 0.0 0.2 \
    --num_crops_per_aug 1 1 \
    --name byol-400ep-custom \
    --entity unitn-mhug \
    --project solo-learn \
    --save_checkpoint \
    --method byol \
    --proj_hidden_dim 4096 \
    --pred_hidden_dim 8192 \
    --base_tau_momentum 0.99 \
    --final_tau_momentum 1.0
```

this is my predict code:

```
import glob
import torch
from PIL import Image
from torchvision import transforms
from torchvision.models import resnet18, resnet50
import torch.nn as nn
from solo.args.setup import parse_args_linear
from solo.methods.base import BaseMethod
from solo.utils.backbones import (
    swin_base,
    swin_large,
    swin_small,
    swin_tiny,
    vit_base,
    vit_large,
    vit_small,
    vit_tiny,
)
from solo.methods.linear import LinearModel


def load_model():
    args = parse_args_linear()

    assert args.backbone in BaseMethod._SUPPORTED_BACKBONES
    backbone_model = {
        "resnet18": resnet18,
        "resnet50": resnet50,
        "vit_tiny": vit_tiny,
        "vit_small": vit_small,
        "vit_base": vit_base,
        "vit_large": vit_large,
        "swin_tiny": swin_tiny,
        "swin_small": swin_small,
        "swin_base": swin_base,
        "swin_large": swin_large,
    }[args.backbone]

    # initialize backbone
    kwargs = args.backbone_args
    cifar = kwargs.pop("cifar", False)
    # swin specific
    if "swin" in args.backbone and cifar:
        kwargs["window_size"] = 4

    backbone = backbone_model(**kwargs)
    if "resnet" in args.backbone:
        # remove fc layer
        backbone.fc = nn.Identity()
        if cifar:
            backbone.conv1 = nn.Conv2d(3, 64, kernel_size=3, stride=1, padding=2, bias=False)
            backbone.maxpool = nn.Identity()

    assert (
        args.pretrained_feature_extractor.endswith(".ckpt")
        or args.pretrained_feature_extractor.endswith(".pth")
        or args.pretrained_feature_extractor.endswith(".pt")
    )
    ckpt_path = args.pretrained_feature_extractor

    state = torch.load(ckpt_path)["state_dict"]
    for k in list(state.keys()):
        if "encoder" in k:
            raise Exception(
                "You are using an older checkpoint."
                "Either use a new one, or convert it by replacing"
                "all 'encoder' occurances in state_dict with 'backbone'"
            )
        if "backbone" in k:
            state[k.replace("backbone.", "")] = state[k]
        del state[k]
    backbone.load_state_dict(state, strict=False)
    del args.backbone
    line_model = LinearModel(backbone, **args.__dict__)
    # line_model.num_classes = 1024
    return line_model


def predict(model, x):
    model.backbone.eval()
    output = model(x)["logits"]
    predicted = int(torch.max(output.data, 1)[1].numpy())
    # predicted = output.argmax(dim=-1).numpy()
    return predicted


def load_one(image_name):
    image = Image.open(image_name).convert("RGB")
    transform_train=transforms.Compose([
                transforms.ToTensor(),
                transforms.Normalize((0.4914, 0.4822, 0.4465), (0.247, 0.243, 0.261)),
    ])
    image = transform_train(image).float()
    image = image.unsqueeze(0)
    return image


if __name__=="__main__":
    m = load_model()
    test_files = glob.glob("/mnt/data/test/cheek/*.jpg")[:50]
    all_img = []
    for f in test_files:
        img=load_one(f)
        output = predict(m, img)
        print(output, ":", f)
```

and my predict bash:

```
python predict.py \
    --dataset custom \
    --backbone resnet18 \
    --data_dir ./ \
    --train_dir ./ \
    --gpus 0 \
    --sync_batchnorm \
    --precision 16 \
    --optimizer sgd \
    --lars \
    --lr 0.1 \
    --weight_decay 1e-5 \
    --batch_size 128 \
    --name general-linear-eval \
    --pretrained_feature_extractor trained_models/byol/offline-w7ez810q/byol-400ep-custom-offline-w7ez810q-ep=399.ckpt \
    --project self-supervised \
```

**Comment by vturrisi** [comment]
Hey @showkeyjar I see a couple of issues. From your predict code alone, you are only loading the backbone, so every time you run your code it will have a differently initialized linear layer hence why your predictions are different. Another issue that I see is that you are doing the pretraining without labels with `--no_labels`. This is not a problem in itself, but you either need to training the linear classifier with our `main_linear.py` afterwards or have the online classifier.

**Comment by showkeyjar** [comment]
e... this makes me more confused. 
why no label trained model need retrain by linear model?
and I found that linear model need validation data?
so, in fact those models need label at last?

but I have nether validation nor labeled data, I'll check result by human, 
what can I do?

**Comment by vturrisi** [comment]
@showkeyjar when you train your model without labelled data, all you get are feature representations. You don't have a classifier to correlate features with specific classes since you never trained it. If you wanna use a classifier, you have a couple of options, namely giving the model labelled data during pretrain (so you can get the online classifier trained) or train a linear classifier on top of the frozen backbone after pretrain (with the `main_linear.py` file). If you don't have any labelled data, the best you can do is use a K-nn classifier (something similar to our `main_knn.py` but there we actually use the labels to compute accuracy and so on).

**Comment by showkeyjar** [comment]
ok, thanks for your reply, I'll try with K-nn classifier.

**Comment by vturrisi** [comment]
Yes, you can set a seed with pytorch/pytorch lightning, but if you are not training the classifier the predictions are not useful. You can also do k-means and check each cluster.
